core/src/main/scala/org/scalafmt/util/TokenOps.scala

```diff
@@ -52,10 +53,14 @@ object TokenOps {
     longHash
   }
 
-  def shouldGet2xNewlines(tok: FormatToken): Boolean = {
+  def shouldGet2xNewlines(tok: FormatToken, style: ScalafmtConfig): Boolean = {
     !isDocstring(tok.left) && {
       val newlines = newlinesBetween(tok.between)
       newlines > 1 || (isDocstring(tok.right) && !tok.left.is[Comment])
+    } || {
+      style.newlines.beforeTopLevelStatements &&
+        tok.between.count(_.is[KwNew]) < 2 &&
+        (tok.right.is[KwObject] || tok.right.is[KwDef] || tok.right.is[KwDef] || tok.right.is[KwCase] || tok.right.is[KwPackage])
```

Comment:
I would think modified top levels (`final`, `sealed`) would also need the newline. 

Also, maybe refactor that into a "isTopLevelStatment" function in FormatOps?